draft-ietf-rats-corim.md

```diff
@@ -2149,11 +2149,21 @@ Reference Value does not match.
 
 ##### Comparison for raw-value entries
 
-> [Andy] *I think this comparison method only works if the entry is at key 4 (because
-there needs to be a mask at key 5). Should we have a Reference Value of this
-which stores `[expect-raw-value raw-value-mask]` in an array?*
+The value stored under `measurement-values-map` key 4 is a `raw-value` entry, which must have type BSTR.
 
-[^issue] https://github.com/ietf-rats-wg/draft-ietf-rats-corim/issues/71
+The value stored under the condition ECT `measurement-values-map` key 5 is a `raw-value-mask` entry, which must have type BSTR.
+
+If the condition ECT does not contain a `raw-values-mask` then the comparison MUST return true if the condition ECT `raw-value` is binary equal to the candidate entry `raw-value`.
+
+If the lengths differ amongst the condition ECT `raw-values-mask`, condition ECT `raw-value` and candidate entry `raw-value`, then the comparison MUST return false.
+
+If all three lengths are the same, then a Verifier MUST iterate over the bits in the `raw-values-mask` which are 1, and compare the values for corresponding bit position in the two `raw-value` fields.
+
+If, for every bit position in `raw-values-mask` whose value is 1, the corresponding bits in both `raw-value`s are equal then the comparison MUST return true.
+
+If there is any bit position in `raw-values-mask` whose value is 1, but the two `raw-value`s differ at the same bit position, then the comparison MUST return false.
+
+Note that if a candidate entry contains a value for `raw-values-mask`, then this does not affect the result of the comparison.
```

Comment:
```suggestion
If the condition ECT has a `raw-value-mask` and no `raw-value`, it is ignored.
If the condition ECT has both a `raw-value-mask`, `raw-value`, then the following conditions must be true for a match:

*  `raw-value` MUST be either a `bstr` or `tagged-bytes`. The compared value of `raw-value` is the `bstr` itself or the tagged `bstr` respectively.
*  the lengths of the condition `raw-value` `bstr`, condition `raw-value-mask`, and the candidate `raw-value` `bstr` are equal.
*  All bit positions in `raw-value-mask` whose value is 1 must imply that the bit values at the same position in both the condition `raw-value` `bstr` and the candidate `raw-value` `bstr` are equal.

Note that the candidate's `raw-value-mask` is ignored if it exists.
```
We can simplify if we want to treat `bstr` and `tagged-bytes` as different since they have different CBOR encodings, and since `bstr` isn't in the base CDDL's `$raw-value-type-choice`, but I figured there would likely be confusion about tagged-bytes vs non-tagged.

The "mask implies raw-value is a bstr" condition is to restrict the behavior of the mask to when the raw-value is truly just bytes that can be masked. Otherwise the type choice's alternatives wouldn't have a clear meaning with the mask's presence.

The quoted text is in 5.1.4.1.4.6 seems to contradict my proposal of limiting the scope to only `bstr`:
> When a new raw value type is defined, the convention for applying the mask is also defined. Typically, a CoRIM profile is used to define new raw values and mask semantics.

I disagree with this. The semantics of a standard codepoint should not be open to interpretation. A profile that wants nonstandard semantics needs to use a nonstandard codepoint.

Reply:
> `raw-value` MUST be either a `bstr` or `tagged-bytes`.

The CDDL says the only `$raw-value-type-choice` defined is `tagged-bytes`. `bstr` only isn't defined.

Reply:
>the lengths of the condition `raw-value` `bstr`, condition `raw-value-mask`, and the candidate `raw-value` `bstr` are equal.

Is the "candidate" the same as the `additions`? If so, it would be more intuitive to say addition or addition candidate.

Reply:
>> When a new raw value type is defined, the convention for applying the mask is also defined. Typically, a CoRIM profile is used to define new raw values and mask semantics.

>I disagree with this. The semantics of a standard codepoint should not be open to interpretation. A profile that wants nonstandard semantics needs to use a nonstandard codepoint.

The extension point in `raw-value` isn't "non-standard" in that the CBOR tagging means there is a specification that defines the semantics of the structure. Nevertheless, the asymmetry of the extensibility seems awkward. Maybe an improvement is a mask type that puts both the value and the mask behind the same CBOR tag? But the convention for extensibility is the CBOR tag definition specifies the convention for representing both the value and the mask as well as the convention for converting the value and mask to a bstr.

Reply:
> > `raw-value` MUST be either a `bstr` or `tagged-bytes`.
> 
> The CDDL says the only `$raw-value-type-choice` defined is `tagged-bytes`. `bstr` only isn't defined.

Yes, I included `bstr` to allow for a profile to specify it as an extension and to avoid footguns if a CoRIM issuer forgets the 560 tag on something that is reasonably expected to always be bytes. There are enough instances of x509 extensions that don't add the ASN.1 bytes tag byte in front of a byte string despite not strictly following the x509 spec for extensions that the value also needs to be an ASN.1 object that I accounted for that here. Might be overly cautious.

> Is the "candidate" the same as the ACS entry? If so, it would be more intuitive to say ACS entry.

The "candidate" word is to reference the induction variable for the iteration over all possible ACS entries. It helps disambiguate which entry we're talking about.

Reply:
I think we need to include non-tagged bstr because some evidence might be in that format. The place where tagging is useful is in raw-values within stateful-environment or similar.